File: test/c/test_permissions.cpp
#include <iostream>
#include <cstdlib>
#include <string>

#include <exception>

#include "ffilesystem.h"

#ifdef _MSC_VER
#include <crtdbg.h>
#endif

int main()
{

#ifdef _MSC_VER
  _CrtSetReportMode(_CRT_ASSERT, _CRTDBG_MODE_FILE);
  _CrtSetReportFile(_CRT_ASSERT, _CRTDBG_FILE_STDERR);
  _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
  _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
  _CrtSetReportMode(_CRT_ERROR, _CRTDBG_MODE_FILE);
  _CrtSetReportFile(_CRT_ERROR, _CRTDBG_FILE_STDERR);
#endif

std::string p = fs_get_permissions("");
if(p.length() != 0)
    throw std::runtime_error("test_exe: get_permissions('') should be empty, got: " + p);

std::string read = "readable.txt", noread = "nonreadable.txt", nowrite = "nonwritable.txt";

fs_touch(read);
fs_set_permissions(read, 1, 0, 0);

p = fs_get_permissions(read);
std::cout << "Permissions for " << read << ": " << p << "\n";

if(p.length() == 0)
    throw std::runtime_error("test_exe: get_permissions('" + read + "') should not be empty");

if(!fs_is_readable(read))
    throw std::runtime_error("test_exe: " + read + " should be readable");

if(!fs_exists(read))
    throw std::runtime_error("test_exe: " + read + " should exist");

if(!fs_is_file(read))
    throw std::runtime_error("test_exe: " + read + " should be a file");

// for Ffilesystem, even non-readable files "exist" and are "is_file"
fs_touch(noread);
fs_set_permissions(noread, -1, 0, 0);

p = fs_get_permissions(noread);
std::cout << "Permissions for " << noread << ": " << p << "\n";

if(p.length() == 0)
    throw std::runtime_error("test_exe: get_permissions('" + noread + "') should not be empty");

if(fs_is_readable(noread)){
    std::cerr << "XFAIL: test_exe: " << noread << " should not be readable\n";
} else {

if(!fs_exists(noread))
    throw std::runtime_error("test_exe: " + noread + " should exist");

if(!fs_is_file(noread))
    throw std::runtime_error("test_exe: " + noread + " should be a file");
}
// writable
if(!fs_is_file(nowrite))
  fs_touch(nowrite);
fs_set_permissions(nowrite, 0, -1, 0);

std::cout << "Permissions for " << nowrite << ": " << fs_get_permissions(nowrite) << "\n";

if(fs_is_writable(nowrite)){
    std::cerr << "ERROR: test_exe: " << nowrite << " should not be writable\n";
    return 77;
}

if(!fs_exists(nowrite))
    throw std::runtime_error("test_exe: " + nowrite + " should exist");

if(!fs_is_file(nowrite))
    throw std::runtime_error("test_exe: " + nowrite + " should be a file");

return EXIT_SUCCESS;
}

File: test/fortran/test_permissions.f90
program main

use, intrinsic :: iso_fortran_env, only : stderr=>error_unit
use filesystem

implicit none

integer :: i
character(9) :: p

character(:), allocatable :: reada, noread, nowrite

allocate(character(get_max_path()) :: reada, noread, nowrite)

if(command_argument_count() < 3) error stop "specify  <readable file> <non-readable file> <non-writable>"

call get_command_argument(1, reada, status=i)
if(i /= 0) error stop "error getting readable file name"

call get_command_argument(2, noread, status=i)
if(i /= 0) error stop "error getting non-readable file name"

call get_command_argument(3, nowrite, status=i)
if(i /= 0) error stop "error getting non-writable file name"

p = get_permissions("")
if(len_trim(p) /= 0) then
    write(stderr, '(a)') "get_permissions('') should be empty: " // p
    error stop
endif

p = get_permissions(reada)
print '(a)', "Permissions for " // trim(reada)// ": "// p

if (len_trim(p) == 0) error stop "get_permissions('"//trim(reada)//"') should not be empty"

if (p(1:1) /= "r") then
    write(stderr, '(a)') "ERROR: test_exe: "//trim(reada)//" should be readable"
    error stop
endif
if(.not. is_readable(reada)) error stop "test_exe: "//trim(reada)//" should be readable"

if(.not. exists(reada)) error stop "test_exe: "//trim(reada)//" should exist"

if(.not. is_file(reada)) error stop "test_exe: "//trim(reada)//" should be a file"

!! for Ffilesystem, even non-readable files "exist" and are "is_file"

p = get_permissions(noread);
print '(a)', "Permissions for " // trim(noread)// ": "// p

if (len_trim(p) == 0) error stop "get_permissions('"//trim(noread)//"') should not be empty"

if (index(p, "r") /= 0) then
    write(stderr, '(a)') "XFAIL:test_exe: "//trim(noread)//" should not be readable"
else
if (is_readable(noread)) error stop "test_exe: "//trim(noread)//" should not be readable"

if (.not. exists(noread)) error stop "test_exe: "//trim(noread)//" should exist"

if (.not. is_file(noread)) error stop "test_exe: "//trim(noread)//" should be a file"
endif

!> writable

if (.not. is_writable(".")) error stop "test_exe: . should be writable"

p = get_permissions(nowrite);
print '(a)', "Permissions for " // trim(nowrite)// ": "// p

if (len_trim(p) == 0) error stop "get_permissions('"//trim(nowrite)//"') should not be empty"

if (index(p, "w") /= 0) then
    write(stderr, '(a)') "test_exe: "//trim(nowrite)//" should not be writable"
    error stop 77
endif
if(is_writable(nowrite)) error stop "test_exe: "//trim(nowrite)//" should not be writable"

if (.not. exists(nowrite)) error stop "test_exe: "//trim(nowrite)//" should exist"

if (.not. is_file(nowrite)) error stop "test_exe: "//trim(nowrite)//" should be a file"

end program
